---
title: "Extended Figure 6"
author: "Christian Ayala Ortiz"
date: "`r Sys.Date()`"
output: html_document
---
# Load libraries

```{r}
library(ggh4x)
library(ggpubr)
library(patchwork)
library(readxl)
library(tidyverse)
source('scripts/custom_functions.R')
```

# Process data

```{r}
metadata <- load_metadata()
mag_taxonomy <- load_mag_taxonomy()

aa_dram <- read_xlsx("data/dram_annotations/metabolism_summary.xlsx", 
                     sheet = 'Organic Nitrogen') %>% 
  select(gene_id, gene_description, module, header) %>% 
  filter(header == 'Amino Acid') %>% 
  mutate(module = str_remove(module, ',.*')) %>% 
  filter(str_detect(module, 'biosynthesis'))

def <- read_csv("data/definitions/definitions_C_N_S_P_11_14_24.csv") %>% 
  filter(Category_2 == 'Phosphate and amino acid transporters') 

annotations <- read_tsv("data/dram_annotations//annotations_merged.tsv")

geTMM_exp <- read_csv("data/expression/geTMM_aminoacids_transporters.csv")

annot_completeness <- aa_dram %>% 
  group_by(module) %>% 
  mutate(n_genes = n()) %>% 
  inner_join(annotations, by = c('gene_id' = 'ko_id')) %>% 
  select(gene_id, module, fasta, n_genes) %>% 
  distinct() %>% 
  group_by(fasta, module, n_genes) %>%
  count() %>% 
  mutate(completeness = n/n_genes) %>% 
  inner_join(mag_taxonomy, by = c('fasta' = 'bin'))

annot_completeness_plot <- annot_completeness %>% 
  ggplot() +
  geom_tile(aes(x = fasta,
                y = module,
                fill = completeness),
            color = 'white') +
  facet_grid2(cols = vars(Phylum),
              scales = 'free',
              space = 'free',
              strip = strip_themed(clip = 'off')) +
  scale_fill_distiller(palette = 'Oranges', direction = 1,
                       labels = scales::percent) +
  custom_theme +
  theme(strip.text.x = element_text(angle = 90),
        panel.spacing = unit(0.05, 'lines'),
        axis.text.x = element_blank(),
        axis.text.y = element_text(size = 4),
        legend.position = 'bottom',
        legend.key.width = unit(.7, 'lines'),
        legend.box.spacing = unit(0.05, 'lines'),
        plot.margin = unit(c(0.05, 0.05, 0.05, 0.05), 'lines'),
        axis.title = element_blank(),
        axis.ticks.x = element_blank())
  
annot_completeness_plot

annot_AA <- annotations %>% 
  select(Gene_id = ...1, fasta, ko_id) %>% 
  inner_join(aa_dram, by = c('ko_id' = 'gene_id')) %>% 
  inner_join(geTMM_exp, by = c('Gene_id', 'fasta', 'ko_id')) %>% 
  group_by(ko_id, fasta, module, header, sampleid) %>% 
  summarise(geTMM = sum(geTMM)) %>% 
  group_by(fasta, module, header, sampleid) %>% 
  summarise(geTMM = mean(geTMM)) %>% 
  inner_join(mag_taxonomy, by = c('fasta' = 'bin')) %>%
  group_by(Phylum, module, header, sampleid) %>% 
  summarise(geTMM = mean(geTMM)) %>% 
  inner_join(metadata, by = c('sampleid' = 'sample')) %>% 
  group_by(Phylum, module, header, month) %>% 
  summarise(geTMM = sum(geTMM)) %>% 
  filter(geTMM > 0)
  
annot_AA_plot <- annot_AA %>% 
  ggplot() +
  geom_tile(aes(x = month,
                y = module,
                fill = log(geTMM)),
            color = 'white') +
  facet_grid2(cols = vars(Phylum),
              strip = strip_themed(clip = 'off')) +
  scale_fill_distiller(palette = 'PuBuGn', direction = 1) +
  custom_theme +
  theme(axis.title = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1),
        strip.text.x = element_text(angle = 90),
        panel.spacing = unit(0.05, 'lines'),
        axis.text = element_text(size = 4),
        legend.position = 'bottom',
        legend.key.width = unit(.7, 'lines'),
        legend.box.spacing = unit(0.05, 'lines'),
        plot.margin = unit(c(0.05, 0.05, 0.05, 0.05), 'lines'))

annot_AA_plot

AA_trans <- annotations %>% 
  select(Gene_id = ...1, fasta, ko_id) %>% 
  inner_join(def, by = c('ko_id' = 'Definition')) %>% 
  inner_join(geTMM_exp, by = c('Gene_id', 'fasta', 'ko_id')) %>% 
  group_by(ko_id, fasta, Category_1, Module_id, sampleid) %>% 
  summarise(geTMM = sum(geTMM)) %>% 
  group_by(fasta, Category_1, Module_id, sampleid) %>% 
  summarise(geTMM = mean(geTMM)) %>% 
  inner_join(mag_taxonomy, by = c('fasta' = 'bin')) %>%
  group_by(Phylum, Category_1, Module_id, sampleid) %>% 
  summarise(geTMM = mean(geTMM)) %>% 
  inner_join(metadata, by = c('sampleid' = 'sample')) %>% 
  group_by(Phylum, Category_1, Module_id, month) %>% 
  summarise(geTMM = sum(geTMM)) %>% 
  filter(geTMM > 0)

trans_AA_plot <- AA_trans %>% 
  ggplot() +
  geom_tile(aes(x = month,
                y = Category_1,
                fill = log(geTMM)),
            color = 'white') +
  facet_grid2(cols = vars(Phylum),
              strip = strip_themed(clip = 'off')) +
  scale_fill_distiller(palette = 'PuBuGn', direction = 1) +
  custom_theme +
  theme(axis.title = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1),
        axis.text = element_text(size = 4),
        strip.text.x = element_text(angle = 90),
        panel.spacing = unit(0.05, 'lines'),
        legend.position = 'bottom',
        legend.key.width = unit(.7, 'lines'),
        legend.box.spacing = unit(0.05, 'lines'),
        plot.margin = unit(c(0.05, 0.05, 0.05, 0.05), 'lines'))

trans_AA_plot

supp_fig8 <- annot_completeness_plot + annot_AA_plot + trans_AA_plot +
  plot_layout(ncol = 1, heights = c(1.5, 1.5, 1)) +
  plot_annotation(tag_levels = 'A')

ggsave('output_figures/Supplementary_Figure_8.svg', supp_fig8, dpi = 300, 
       height = 185, width = 180, units = 'mm')
ggsave('output_figures/Supplementary_Figure_8.png', supp_fig8, dpi = 300, 
       height = 185, width = 180, units = 'mm')
```


